### PR TITLE
Sort signatures by keys IDs for deterministic signature order

### PR DESCRIPTION
**Pull request by ethan-lowman-dd**
In the course of implementing delegations in the repo writer ([WIP branch](https://github.com/ethan-lowman-dd/go-tuf/pull/1/files)), I had to change how keys are stored internally, which introduced non-determinism in the order of signatures. This caused tests which relied on fixtures to become flaky, as the order of signatures was not consistent. In order to keep support for static fixtures (using deterministic signature schemes like Ed25519), this PR sorts signatures by key ID.

### CODE REVIEW

File: repo.go [hunk 1]
```diff
@@ -605,7 +607,10 @@ func (r *Repo) getSigningKeys(name string) ([]sign.Signer, error) {
 		return nil, err
 	}
 	if name == "root" {
-		return signingKeys, nil
+		sorted := make([]sign.Signer, len(signingKeys))
+		copy(sorted, signingKeys)
+		sort.Sort(signer.ByIDs(sorted))
```
Comment (trishankatdatadog) [inline repo.go:612]: This line is a bit hard to parse. I am sort sorting signer which are already sorted?

File: repo_test.go [hunk 2]
```diff
@@ -543,9 +552,17 @@ func (rs *RepoSuite) TestSign(c *C) {
...
+		// Signatures may be in any order, so must sort key IDs before comparison.
+		wantKeyIDs := append([]string{}, keyIDs...)
+		sort.Strings(wantKeyIDs)
+
+		gotKeyIDs := []string{}
+		for _, sig := range s.Signatures {
+			gotKeyIDs = append(gotKeyIDs, sig.KeyID)
 		}
+
+		c.Assert(sortStrings(keyIDs), DeepEquals, sortStrings(gotKeyIDs))
```
Comment (trishankatdatadog) [inline repo_test.go:565]: What are we really testing here? That we sorted strings correctly? So why not just reuse the built-in function for sorting strings in the first place?

**Comment by rdimitrov** [comment]
Apart from #143, it's also worth syncing with @asraa because of #148 as both of the PRs alter/depend on the same `sign` package.

In that sense, maybe it would be better to prioritize which of these should be merged first to avoid the burden of fixing conflicts afterwards?

File: internal/signer/sort.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	iIDs := make([]string, len(ids))
+	copy(iIDs, ids)
+	sort.Strings(iIDs)
+
+	ids = b[j].IDs()
+	jIDs := make([]string, len(ids))
+	copy(jIDs, ids)
+	sort.Strings(jIDs)
+
+	minLen := len(iIDs)
```
[the remark this answers is not in the document]
Comment (ethan-lowman-dd) [reply internal/signer/sort.go:33]: Go doesn't have a min function for two ints, so these four lines (33-36) are the best we can do.

File: repo_test.go (hunk 2 above)
Comment (ethan-lowman-dd) [reply repo_test.go]: This tests that keyIDs and gotKeyIDs have the same elements, disregarding order. Previously the test only passed if the key IDs were in the same order.

I had a duplicate sorting of `keyIDs`/`wantKeyIDs`, which was perhaps a cause of confusion -- I've fixed this.

The `sortStrings` helper made a copy of the string slice before sorting (since `sort.Strings` sorts in place), but I removed the helper since it was confusing and only used in this one test.

File: repo.go (hunk 1 above)
Comment (ethan-lowman-dd) [reply repo.go:612]: Lines 610-611 makes a copy of `signingKeys` called `sorted` (which is not yet sorted until the next line). Line 612 sorts `sorted` in place using the `signer.ByIDs` comparator.

**Review by asraa — COMMENTED** [review]
Thanks! I'm not opposed to this change, but could you point me to where the testdata needed to be ordered?

I'm seeing a few checks in the interop_test/client_test but they don't seem to require ordering.

Just wondering if it makes more sense to change tests to test equality/expectation without ordering

File: repo_test.go [hunk 4]
```diff
@@ -543,9 +546,18 @@ func (rs *RepoSuite) TestSign(c *C) {
...
+		wantKeyIDs := append([]string{}, keyIDs...)
+		sort.Strings(wantKeyIDs)
+
+		gotKeyIDs := []string{}
+		for _, sig := range s.Signatures {
+			gotKeyIDs = append(gotKeyIDs, sig.KeyID)
 		}
+		sort.Strings(gotKeyIDs)
+
+		c.Assert(wantKeyIDs, DeepEquals, gotKeyIDs)
```
Comment (asraa) [inline repo_test.go:560]: client_test approaches this with https://github.com/theupdateframework/go-tuf/blob/f38e930df33905abc5bb7be5fcd0668231807a88/util/util.go#L261

**Comment by ethan-lowman-dd** [comment]
@asraa The tests that are flaky if signature order is nondeterministic are:

1. File hash checks against generated fixtures:
    * https://github.com/theupdateframework/go-tuf/blob/master/client/interop_test.go#L41-L48
2. Checking key IDs in order:
    * https://github.com/theupdateframework/go-tuf/blob/master/repo_test.go#L545-L548
    * https://github.com/theupdateframework/go-tuf/blob/master/repo_test.go#L1181-L1184

I addressed (1) by sorting signatures by key IDs and regenerating these fixtures. I addressed (2) by sorting keys IDs in the tests before making comparisons.

File: repo_test.go (hunk 4 above)
Comment (ethan-lowman-dd) [reply repo_test.go:560]: This isn't quite the same as comparing sorted lists, right? For example:
```
a := []string{"AAA", "BBB", "CCC"}
b := []string{"AAA, "BBB", "CCC", "CCC"}

aSet := StringSliceToSet(a) 
bSet := StringSliceToSet(b)

// This evaluates to true, but the slices we're comparing are not equal.
reflect.DeepEqual(aSet, bSet)
```

I think in this case, we would want tests to fail if signatures are erroneously duplicated. Comparing sorted lists accomplishes this, but comparing sets does not.

Comment (asraa) [reply repo_test.go:560]: good point with duplicates. i'll have to look into that if that test makes sense to me now too

File: internal/signer/sort.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package signer
```
[the remark this answers is not in the document]
Comment (ethan-lowman-dd) [reply internal/signer/sort.go:1]: Added tests

File: repo.go [hunk 6]
```diff
@@ -605,7 +607,10 @@ func (r *Repo) getSigningKeys(name string) ([]sign.Signer, error) {
 		return nil, err
```
[the remark this answers is not in the document]
Comment (ethan-lowman-dd) [reply repo.go]: With your other comment, this method no longer sorts the signers.

[the remark this answers is not in the document]
Comment (ethan-lowman-dd) [reply repo.go]: I updated the PR to sort on usage rather than on retrieval. I'm not too opinionated about which is better.

[the remark this answers is not in the document]
Comment (ethan-lowman-dd) [reply repo.go:607]: As discussed, I reverted this change and renamed the method to `getSortedSigningKeys`.

**Review by trishankatdatadog — DISMISSED** [review]

**Review by asraa — DISMISSED** [review]

**Review by hosseinsia — APPROVED** [review]
LGTM!

File: internal/signer/sort_test.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+		value: json.RawMessage(`{"mock": 5}`),
+	}
+
+	s := []keys.Signer{
+		s1, s2, s3, s4, s5,
+	}
+
+	sort.Sort(signer.ByIDs(s))
+
+	sorted := []keys.Signer{
```
Comment (hosseinsia) [inline internal/signer/sort_test.go:64]: Sorry I did not realize this part last time. I think the reason for this random order is because the IDs is the hash of the value? This makes this test very unreadable. 
Can you please mock the IDs in the PublicKey to return the exact value (instead of the hash of the value)? 
and in line 59 input a random order of the signers ... in line 65 change the variable to expected and orders to be s1, s2, s3, s4, s5?

Comment (hosseinsia) [reply internal/signer/sort_test.go]: Hmmm I think that can not quite address the issue.

**Review by trishankatdatadog — APPROVED** [review]
